# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLQuery.cpp
# Commit: 71c122ac0ca7
# Full Hash: 71c122ac0ca73391866b1ef19f4f82bc2d28568b
# Author: David Parks <davidp99@gmail.com>
# Date: 2020-01-09 03:40:18
# Regressor Bug: 1477756
# File Overlap Count: 1
# Description:
#   Bug 1477756 - Initial out-of-process WebGL implementation. r=mccr8,handyman
#   
#   Splits WebGLContext into ClientWebGLContext and HostWebGLContext.  The Client enables the JS-control of a WebGL context in a content procecss while the Host executes the WebGL graphics operations (via a WebGLContext that maintains much of the existing code) in the compositor process.  At this point, the cross-process behavior is disabled -- this series of patches is an incremental step toward that final goal.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D54018
# ==============================================================================

diff -r 5cbf5bb11c58 -r 71c122ac0ca7 dom/canvas/WebGLQuery.cpp
--- a/dom/canvas/WebGLQuery.cpp	Wed Jan 08 21:52:03 2020 +0000
+++ b/dom/canvas/WebGLQuery.cpp	Wed Jan 08 22:19:14 2020 +0000
@@ -91,8 +91,7 @@
   availRunnable->mQueries.push_back(this);
 }
 
-void WebGLQuery::GetQueryParameter(GLenum pname,
-                                   JS::MutableHandleValue retval) const {
+MaybeWebGLVariant WebGLQuery::GetQueryParameter(GLenum pname) const {
   switch (pname) {
     case LOCAL_GL_QUERY_RESULT_AVAILABLE:
     case LOCAL_GL_QUERY_RESULT:
@@ -100,16 +99,18 @@
 
     default:
       mContext->ErrorInvalidEnumInfo("pname", pname);
-      return;
+      return Nothing();
   }
 
   if (!mTarget) {
     mContext->ErrorInvalidOperation("Query has never been active.");
-    return;
+    return Nothing();
   }
 
-  if (mActiveSlot)
-    return mContext->ErrorInvalidOperation("Query is still active.");
+  if (mActiveSlot) {
+    mContext->ErrorInvalidOperation("Query is still active.");
+    return Nothing();
+  }
 
   // End of validation
   ////
@@ -119,9 +120,9 @@
       (mCanBeAvailable || StaticPrefs::webgl_allow_immediate_queries());
   if (!canBeAvailable) {
     if (pname == LOCAL_GL_QUERY_RESULT_AVAILABLE) {
-      retval.set(JS::BooleanValue(false));
+      return AsSomeVariant(false);
     }
-    return;
+    return Nothing();
   }
 
   const auto& gl = mContext->gl;
@@ -130,8 +131,7 @@
   switch (pname) {
     case LOCAL_GL_QUERY_RESULT_AVAILABLE:
       gl->fGetQueryObjectuiv(mGLName, pname, (GLuint*)&val);
-      retval.set(JS::BooleanValue(bool(val)));
-      return;
+      return AsSomeVariant(static_cast<bool>(val));
 
     case LOCAL_GL_QUERY_RESULT:
       switch (mTarget) {
@@ -154,13 +154,9 @@
         case LOCAL_GL_TRANSFORM_FEEDBACK_PRIMITIVES_WRITTEN:
         case LOCAL_GL_TIME_ELAPSED_EXT:
         case LOCAL_GL_TIMESTAMP_EXT:
-          retval.set(JS::NumberValue(val));
-          break;
-
-        default:
-          MOZ_CRASH("Bad `mTarget`.");
+          return AsSomeVariant(val);
       }
-      return;
+      MOZ_CRASH("Bad `mTarget`.");
 
     default:
       MOZ_CRASH("Bad `pname`.");
@@ -206,16 +202,4 @@
   availRunnable->mQueries.push_back(this);
 }
 
-////
-
-JSObject* WebGLQuery::WrapObject(JSContext* cx,
-                                 JS::Handle<JSObject*> givenProto) {
-  return dom::WebGLQuery_Binding::Wrap(cx, this, givenProto);
-}
-
-NS_IMPL_CYCLE_COLLECTION_WRAPPERCACHE_0(WebGLQuery)
-
-NS_IMPL_CYCLE_COLLECTION_ROOT_NATIVE(WebGLQuery, AddRef)
-NS_IMPL_CYCLE_COLLECTION_UNROOT_NATIVE(WebGLQuery, Release)
-
 }  // namespace mozilla